ckanext/security/fanstatic/login_ajax.js

```diff
@@ -48,6 +48,16 @@
         showMfaForm(loginState)
       },
       error: function (xhr) {
+        var jsonResponse = null;
+        try {
+          jsonResponse = JSON.parse(xhr.responseText)
+        } catch (e) {}
+
+        if (jsonResponse && jsonResponse.hasOwnProperty('mfaCodeValid')) {
```

Comment:
This can only occur if no exception was thrown, yes? So it could be moved inside the `try` block.

Reply:
Yep, could be moved into the try catch.

The [hasOwnProperty](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty) function is used in this case to check if the property exists on the response object, as it has different structures depending on the error. I don't really care about the value that the property points to, just that response has that property on it.

Reply:
It's odd, though, having a property named _valid_ and failing as a result.

Reply:
Updated to move the checks inside the try/catch and the wrapped it all under the 403 check, based on the controller response patterns. Also updated the check to be `if (jsonResponse && jsonResponse.mfaCodeValid === false) {`

Local testing shows that this logic works correctly and makes more sense.